project file unification take 2

baronfel opened the pull request:
Ok lets try this again. Smooshed the projects together again, fixed up some dependency shenanigans, did some conditional includes where appropriate, and added some type annotations so that things would build on recent mono. Integration tests are green on my box, let's see what CI thinks.

baronfel committed:
move the projects over to netcore-and-netframework

baronfel committed:
update build script to build different configurations to different output diectories

baronfel committed:
fix suave update

baronfel committed:
tighten frameworks

baronfel committed:
fixup some more deps

baronfel committed:
fix file path

baronfel committed:
fix  build breaks due to api surface area differences

baronfel committed:
don't publish for windows for net461, that's just weird

baronfel committed:
update sdk versions for travis

baronfel committed:
publish correct netcore project

baronfel committed:
fix travis by updating build agent

baronfel commented:
Green! How does this look to everyone?

cartermp commented on build.fsx:
```diff
@@ -75,7 +63,7 @@ let (|AnyNetcoreRuntime|_|) r =
   | FSACRuntime.NET -> None
 
-let isTestSkipped cfg fn =
+let isTestSkipped cfg (fn: string) =
```
Spanatee the manatee says hi :)

baronfel replied:
tell me about it :D  I hit this on every project these days. and each new mono upgrade is a game of 'discover the new overloads!'

cartermp commented on src/FsAutoComplete/FsAutoComplete.fsproj:
```diff
@@ -2,12 +2,15 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net461</TargetFramework>
+    <TargetFrameworks>net461;netcoreapp2.0</TargetFrameworks>
```
Is .NET Framework needed as a TFM still?

baronfel replied:
I'd like to keep it because it includes the old ProjectCrackerVerbose, which is a wrapper around the .net-framework only project cracker, which may be better-supported/more-compatible with old-style project files.  If/when we drop it, though, we could definitely standalone-publish FSAC and have a xcopy-deployable standalone version which would be pretty neat.

Krzysztof-Cieslak replied:
>Is .NET Framework needed as a TFM still?

As far as I know, Ionide is only editor using Core version of the FSAC.

> which may be better-supported/more-compatible with old-style project files

I definitely had some issues where new project parser fails for an odd old-style project, while FCS' ProjectCrapper works just fine... I think it's even happening on the VF# repo.

enricosada replied:
Main issue as @Krzysztof-Cieslak said is the different project loader atm.
plus vim/emacs only use .net version (but that's easy to change in these extensions)

.NET is needed because there the `projectcracker` is used to load the old fsproj (verbose fsproj) loading the msbuild lib too.

With the `dotnet-proj-info` project parser, usually the fsproj loading fails if the msbuild on PATH is wrong (usually wrong visual studio command prompt). i was able to load the VF# sln in ionide.

I am addressing that in `dotnet-proj-info` repo, i hope to have a new release soon, refacotring out the project loading from FSAC and embedding the discovery and choice of msbuild to use in `dotnet-proj-info`.

We cannot use the new `dotnet-proj` on .net FSASC in parallel (with a feature flag), because it will conflict with the msbuild libs loaded by projectcracker

Krzysztof-Cieslak commented:
@baronfel, can you try enabling also .Net Core tests?

enricosada commented on global.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "sdk": {
+        "version": "2.2.100"
```
can we use latest .NET Core Sdk LTS instead of `Current`? no need to go edge
so `2.1.500`

cartermp replied:
Current will likely give some perf benefits. I'd only recommend LTS if there's no intention to bump the version of FSAC except for every year or so.

enricosada replied:
we target `netcoreapp2.0` in FSAC, and the .net core version published as FDD (like in ionide) will run with the installed .net core runtime.

So will use the installed .net core runtime anyway, no perf gain at runtime.
The compiler is faster on .net core 2.2 than 2.1?

the sdk is ok to be stable and slow moving, less bugs more workarounds.

Maybe we can bump to `netcoreapp2.1` later too btw (with a note on minimal requirements).

enricosada reviewed (changes requested):
i see

```
E:\github\fsharp\FsAutoComplete\src\FsAutoComplete\FsAutoComplete.fsproj : warning NU1701: Package 'FSharp.Compiler.Service.ProjectCracker 25.0.1' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETCoreApp,Version=v2.0'. This package may not be fully compatible with your project. [E:\github\fsharp\FsAutoComplete\FsAutoComplete.sln]
```

that shouldnt happen, projectcracker it shouldnt be added to the .net core version at all, otherwise pull msbuild assemblies

i checked and it add `FSharp.Compiler.Service.ProjectCracker` to references

seems to be pulled in by `FSharpLint.Core 0.9.0-beta` who has wrong deps, fixed version is `0.9.0`

ghost commented:
On the project build side, I beleive that a mono paket is currently distributed in `.paket`, this means that a netcore only build will fail (`dotnet build -f netcoreapp2.0`). We now have paket as a dotnet cli tool, (on my machine I can just call `paket restore`), so it should be possible to preference that and fall back to mono if it doesn't exist?

```
$ dotnet build -f netcoreapp2.0
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  /bin/sh: 2: /tmp/tmpbe2242f8ef15483fbb04c8dfed7fde4f.exec.cmd: mono: not found
/home/user/extra/FsAutoComplete/.paket/Paket.Restore.targets(111,5): error MSB3073: The command "mono --runtime=v4.0.30319 "/home/user/extra/FsAutoComplete/.paket/paket.exe" restore" exited with code 127. [/home/user/extra/FsAutoComplete/src/FsAutoComplete.SymbolCache/FsAutoComplete.SymbolCache.fsproj]
```

enricosada commented:
@abk-x yes, but afaik paket shouldnt be installed as global (`-g`), otherwise is not possible to use a specific version in the repo (and can be not in sync with the `paket.restore.target` or at least behave differently than ci)

ghost commented:
I got the build working by doing `rm -rf .paket && paket install` then `dotnet build -f netstandard2.0 src/FsAutoComplete/FsAutoComplete.fsproj`

enricosada committed:
paket ignore the test projects

enricosada committed:
bump to use the .NET Core Sdk LTS instead of Current

enricosada committed:
remove ProjectCracker from `netcoreapp2.0` deps

the FSharpLint.Core v0.9.0-beta had a bad nuspec, the v0.9.0 fix that

enricosada committed:
ionide ignore the test assets

enricosada committed:
migrate test proj to .net sdk

enricosada committed:
add test proj to sln

enricosada committed:
reenable .net core tests stdio

enricosada committed:
fix travis .net core sdk requirement

enricosada committed:
fix paths for .net core tests

enricosada commented:
linter exception:

```
  "Data": "Something went wrong during linter: System.TypeInitializationException: The type initializer for 'FSharpLint.Framework.Configuration' threw an exception. 
---> System.TypeInitializationException: The type initializer for '<StartupCode$FSharpLint-Core>.$Configuration' threw an exception. 
---> System.Exception: Resource 'DefaultConfiguration.FSharpLint' not found in assembly 'FSharpLint.Core, Version=0.9.0.0, Culture=neutral, PublicKeyToken=null'\n   at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1645.Invoke(String message)\n   at <StartupCode$FSharpLint-Core>.$Configuration..cctor()\n   
--- End of inner exception stack trace 
---\n   at FSharpLint.Framework.Configuration..cctor()\n   
--- End of inner exception stack trace ---\n   at FSharpLint.Application.ConfigurationManager.ConfigurationManager.GetConfigurationForProject(String projectFilePath)\n   at <StartupCode$FsAutoComplete-Core>.$Commands.Lint@544.Invoke(Unit unitVar) in E:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/Commands.fs:line 558"
```

because the fsharplint v0.9.0 has the wrong resource name

![image](https://user-images.githubusercontent.com/147243/51287226-cd635880-19f6-11e9-9fbf-438ff7c96485.png)

enricosada committed:
storage:none for Tests packages too

enricosada committed:
bump fsharplint but i think has a regression if file has no module

see output files diff for the regression, but it isnt a common scenario, should be
addressed later

cannot try v0.10.2 because it require dotnet.projinfo v0.20 and that
has a breaking change i dont want to address now

Krzysztof-Cieslak commented:
I’ve bumped Linter also when working on #320

enricosada commented:
@Krzysztof-Cieslak 	nice, let's use v0.10.1 for now, like LSP PR is ok. Do not use v0.10.2 because depends on a  newer version of `dotnet.projinfo` who has a breaking change i will address later in FSAC.
The version 0.20.0 add also a nice surprise, the helper `.target` file in `obj` dir is not needed anymore :D, so zero garbage when loading projects and no interference with incremental compilation

enricosada commented:
as a note, running most of the .net core FSAC tests in command prompt raise this error

```
  "Data": "Request cancelled (exn was System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified\n
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)\n
   at System.Diagnostics.Process.Start()\n   at FsAutoComplete.Utils.runProcess(FSharpFunc`2 log, String workingDir, String exePath, String args) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/Utils.fs:line 531\n
   at FsAutoComplete.NETFrameworkInfoProvider.runCmd@51.Invoke(String exePath, IEnumerable`1 args) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 51\n
   at FsAutoComplete.NETFrameworkInfoProvider.msbuildExec@56-1.Invoke(FSharpList`1 arg10) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 56\n
   at Dotnet.ProjInfo.Inspect.msbuild[a](MSBuildExePath msbuildExe, FSharpFunc`2 run, String project, FSharpList`1 args)\n
   at FsAutoComplete.NETFrameworkInfoProvider.msbuildExec@56-2.Invoke(String project, FSharpList`1 args) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 56\n
   at Dotnet.ProjInfo.Inspect.getProjectInfoOldSdk[a,b,c,d](FSharpFunc`2 log, FSharpFunc`2 msbuildExec, FSharpFunc`2 getArgs, FSharpList`1 additionalArgs, a projPath)\n
   at FsAutoComplete.NETFrameworkInfoProvider.getInfoFromMsbuild[a](FSharpFunc`2 getArgs, FSharpList`1 additionalProps) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 61\n
   at FsAutoComplete.NETFrameworkInfoProvider.getInstalledNETVersions() in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 90\n
   at FsAutoComplete.NETFrameworkInfoProvider.installedNETVersionsLazy@97.Invoke(Unit unitVar) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 97\n
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)\n
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)\n
   at System.Lazy`1.CreateValue()\n
   at FsAutoComplete.NETFrameworkInfoProvider.installedNETVersions() in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/NETFrameworkInfoProvider.fs:line 99\n
   at FsAutoComplete.Environment.netReferecesAssembliesTFM() in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/Environment.fs:line 136\n
   at FsAutoComplete.Environment.netReferecesAssembliesTFMLatest() in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/Environment.fs:line 144\n
   at <StartupCode$FsAutoComplete-Core>.$CompilerServiceInterface.GetProjectOptionsFromScript@511.Invoke(Unit unitVar) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/CompilerServiceInterface.fs:line 511\n
   at Microsoft.FSharp.Control.AsyncPrimitives.CallThenInvoke[T,TResult](AsyncActivation`1 ctxt, TResult result1, FSharpFunc`2 part2)\n
   at <StartupCode$FsAutoComplete-Core>.$Commands.Parse@269-2.Invoke(AsyncActivation`1 ctxt) in e:/github/fsharp/FsAutoComplete/src/FsAutoComplete.Core/Commands.fs:line 269\n   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction))"

```

running the same in VS Commnad proimpt is ok.

Seems the .NET Core FSAC version crash if msbuild is not in PATH, but shouldnt.
maybe that lazy load of assemblies paths shouldnt happen unless needed, or at least be an  `Option` to not crash

Test in CI doesnt show the issue, because mono msbuild or VS msbuild is already in PATH

related to https://github.com/ionide/ionide-vscode-fsharp/issues/656 probably

enricosada committed:
the fsac .net core load a different version of FSharp.Core for fsx

the .NET FSAC load the FSharp.Core used by FSAC and in the tests that's from `bin/net461` who doesnt contains the xmldoc -> no additional info

the .NET Core FSAC instead resolve correctly the path from GAC, and that contains the xmldoc

enricosada commented:
linter on fsac .NET fails for a strong naming issue of fsharplint and fparseccs

```
  "Kind": "info",
  "Data": "Something went wrong during linter: System.TypeInitializationException: The type initializer for 'FSharpLint.Framework.Configuration' threw an exception. ---> System.TypeInitializationException: The type initializer for '<StartupCode$FSharpLint-Core>.$Configuration' threw an exception. ---> System.IO.FileLoadException: Could not load file or assembly 'FParsecCS, Version=1.0.3.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)\n   at FSharpLint.Framework.Configuration.parseHint@261(String hint)\n   at FSharpLint.Framework.Configuration.parseHints@270.Invoke(String x)\n   at Microsoft.FSharp.Primitives.Basics.List.map[T,TResult](FSharpFunc`2 mapping, FSharpList`1 x)\n   at FSharpLint.Framework.Configuration.parseHints(XElement el)\n   at FSharpLint.Framework.Configuration.parseSetting(XElement setting)\n   at FSharpLint.Framework.Configuration.analyserDetails@309.Invoke(XElement settingElement)\n   at Microsoft.FSharp.Collections.Internal.IEnumerator.map@74.DoMoveNext(b& curr)\n   at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext()\n   at Microsoft.FSharp.Collections.MapTreeModule.mkFromEnumerator[a,b](IComparer`1 comparer, MapTree`2 acc, IEnumerator`1 e)\n   at Microsoft.FSharp.Collections.MapTreeModule.ofSeq[TKey,T](IComparer`1 comparer, IEnumerable`1 c)\n   at Microsoft.FSharp.Collections.FSharpMap`2.Create(IEnumerable`1 ie)\n   at FSharpLint.Framework.Configuration.parseAnalyser(XElement analyser)\n   at FSharpLint.Framework.Configuration.configuration@326.Invoke(XElement analyser)\n   at Microsoft.FSharp.Collections.Internal.IEnumerator.map@74.DoMoveNext(b& curr)\n   at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext()\n   at Microsoft.FSharp.Collections.MapTreeModule.mkFromEnumerator[a,b](IComparer`1 comparer, MapTree`2 acc, IEnumerator`1 e)\n   at Microsoft.FSharp.Collections.MapTreeModule.ofSeq[TKey,T](IComparer`1 comparer, IEnumerable`1 c)\n   at Microsoft.FSharp.Collections.FSharpMap`2.Create(IEnumerable`1 ie)\n   at FSharpLint.Framework.Configuration.configuration(String file)\n   at <StartupCode$FSharpLint-Core>.$Configuration..cctor()\n   --- End of inner exception stack trace ---\n   at FSharpLint.Framework.Configuration..cctor()\n   --- End of inner exception stack trace ---\n   at FSharpLint.Framework.Configuration.get_defaultConfiguration()\n   at FSharpLint.Application.ConfigurationManager.ConfigurationManager.GetConfigurationForProject(String projectFilePath)\n   at <StartupCode$FsAutoComplete-Core>.$Commands.Lint@544.Invoke(Unit unitVar)"
```

**disabled the test for now, but the issue should be fixed otherwise linter fails on FSAC .NET**

enricosada committed:
known failure of lint on .NET FSAC

enricosada committed:
log test failure to output.json

ghost commented:
does https://github.com/fsharp/FsAutoComplete/pull/325#issuecomment-455001979 relate to https://github.com/ionide/ionide-vscode-fsharp/issues/924#issuecomment-454999848 ? 

Edit: It looks very similar to what I have run into when I have attempted to use FSAC without mono installed

enricosada commented:
@abk-x no afaik the comment is for the issue loading fsx.
To load the intellisense of an fsx, you need the paths for System, System.Core etc.
The fsac .net core ask msbuild these info. But if `msbuild` is not in PATH, fsac crash with that stupid error, instead of saying "msbuild not found in PATH" (atm fsac cannot use customized msbuild paths, so try `msbuild` in PATH, that's why there is https://github.com/fsharp/FsAutoComplete/pull/251 )
As a note, because fsx are for fsi, i expect to be for .NET (not .net core), so i ask to `msbuild` not `dotnet msbuild`.
For some scenarios using fsx with .net core only may be ok (fake for example), and should be enough to ask `dotnet msbuild`, but again, the msbuild host to use is not configurable atm in FSAC or ionide.

Anyway that's not related to the issue https://github.com/ionide/ionide-vscode-fsharp/issues/924 about loading fsproj.
That's ihmo an issue of timing of loading the project, if fails but it's a false positive and bad info without compiler args (the `OtherOptions` so intellisense doesnt work) are cached and never invalidated.

I seen your comment and the mention. i'll try tomorrow (a bit tired atm to repro) and comment it.

enricosada commented:
status:

there is a .net core integration test failing ( https://ci.appveyor.com/project/rneatherway/fsautocomplete/builds/21679761#L585 ).
dunno the reason, i cannot repro locally.
tomorrow i'll try to RDP into appveyor and see, rerunning the same `fsi` will show the error probably

Krzysztof-Cieslak commented:
@enricosada, It's probably the discussion for a different issue but... Why do we have `msbuild` in `PATH` dependency for FSAC-netcore? As far as I can tell our `msbuild` detection (https://github.com/fsharp/FsAutoComplete/blob/master/src/FsAutoComplete.Core/Environment.fs#L63-L84) is accurate, so can't we just use the path we find with it?

enricosada commented:
@Krzysztof-Cieslak because it was the only way atm to choose what msbuild to use, if multiple are installed. the fsproj need to be built with the right msbuild.
Atm we cannot choose it (that's why the PR #251 ) 
The user can just open code from the correct vs command prompt. i use it like this a lot at work

but yes seems it is too fragile, so if not found in the PATH it should fallback to that detection and use the latest.

enricosada committed:
enable rdp

enricosada committed:
empty commit

enricosada committed:
lower timeout

enricosada committed:
Revert "enable rdp"

This reverts commit 81c68fb0ea1c9f38748135f6b6c6a4f72e91d2ef.

enricosada committed:
readd remove doc

enricosada committed:
by default, skip msbuild 4 tests

enricosada committed:
x64 exe

enricosada committed:
show fsi exception

enricosada committed:
doc

enricosada committed:
skip msbuild 4 tests if not installed

for .net fsac, mean msbuild assemblies not in gac

enricosada committed:
f

enricosada committed:
another msbuild 4

enricosada committed:
avoid deadlock if code throw an exception

enricosada commented:
some test fails because fsac goes in deadlock, waiting for `Project.Response` in [Commands.fs#L339](https://github.com/fsharp/FsAutoComplete/blob/b09784a96de2d0fe59409896fc8263a5168b1805/src/FsAutoComplete.Core/Commands.fs#L339)

```fsharp
    member __.Response with get() = agent.PostAndReply GetResponse
```

i think that `PostandReply` deadlock, because the mailboxprocess raise an unhandled exception

the ci job show the failed test, but doesnt end.
that's another thing, dunno if that fake test runner handle well a dead locked fsautocomplete

enricosada committed:
check if project file exists before get info

enricosada committed:
run ci using recent mono

enricosada committed:
get msbuild path from FSAC, do not expect `msbuild` in PATH

enricosada commented:
@Krzysztof-Cieslak trying to use the msbuild of FSAC with c7a8917 instead of the one in PATH

enricosada committed:
fix build

enricosada committed:
fix build

baronfel commented:
Way to bring it home, @enricosada!

enricosada commented:
Ok merged, reenabled .net core std integration tests too.

Awesome work @baronfel 👍

atm linter on .NET FSAC doesnt work, see info in commit message
Let's continue work from `master` to fix disabled tests and maybe reenable http tests too.
